### PR TITLE
Standalone terminfo definition

### LINKED ISSUE #1489: Create a standalone terminfo definition
In #1485, it was suggested that the current alacritty terminfo definition, which is based around the system's `xterm-256color` definition, be replaced with a custom definition without any system dependencies. I.e. make it completely standalone.

For a complete background, see the following comments:

* https://github.com/jwilm/alacritty/issues/1485#issuecomment-409659033
* https://github.com/jwilm/alacritty/issues/1485#issuecomment-409673655
* https://github.com/jwilm/alacritty/issues/1485#issuecomment-409719745

The idea would be to start out with an xterm clone. That way, there will be no functional differences from the current definition.

### PR DESCRIPTION
**Pull request by dnkl**
This replaces the current alacritty terminfo definitions with three standalone variants; `alacritty`, `alacritty-256color` and `alacritty-direct`.

They are all completely standalone, meaning they do not use any definitions from the system's terminfo database.

They are based off of the corresponding xterm definitions from ncurses 6.1. I've kept the fragments as-is. This makes it easier to compare with the xterm definitions (and to pull in future changes/fixes from xterm). It does however make the file much more verbose and perhaps unnecessarily difficult to navigate. It might make sense to merge all the basic fragments into the `alacritty` definition.

I've removed all the fragment comments, since I felt they apply to xterm but not alacritty.

`infocmp` shows no differences between the new alacritty definitions and xterm's. I'm putting it up for others to try it out. @chrisduerr .

Closes #1489 

- [x] Remove fragmentation of terminfo
- [x] Rename alacritty-256color to alacritty
- [x] Set the default $TERM used by Alacritty to alacritty
- [x] ~Remove export TERM="alacritty" part from README~ (couldn't find anything like this anywhere)
- [x] Update documentation of TERM in alacritty.yml
- [x] ~Clean up/remove capabilities tic complains about.~
- [x] Update README.md

### CODE REVIEW

**Comment by dnkl** [comment]
If you don't want to overwrite you user's regular terminfo database (or the system database), you can do `tic -o/tmp alacritty.info`. And then set the `TERMINFO` environment variable to, in this case, `/tmp`. E.g. `TERMINFO=/tmp infocmp xterm-256color alacritty-256color`.

**Comment by dnkl** [comment]
> Instead of copying all definitions available for xterm, I'd recommend just copying all definitions required for Alacritty to work properly with all available features.

I agree. I thought it best to start out this way though, to get a clean start. Going forward, there are two ways in which we can strip away definitions:

1. Get rid of unused/unsupported capabilities. For example, `tic` already complains about a large number of "unknown" capabilities (unclear why, and I should probably find out why). But more importantly, remove those not supported and/or required by alacritty. **I'll likely need help with this**.

2. Remove e.g. the `alacritty` definition and perhaps rename the `alacritty-256color` definition to `alacritty`. I doubt there are any reasons to keep the non-256 version. In terms of LOCs, this however will only remove about 2-3 lines...

I think it would be a good idea to keep the `direct` version separate from the regular 256 color version, at least for now. At least on arch, `ncmpcpp` crashes when using a `TERM=*-direct`. Not just under alacritty, but under any terminal emulator.

> Looking at other terminal emulators, rxvt-unicode provides only urxvt and urxvt-256color as available terminfo files for example. Kitty goes one step further and just provides kitty with nothing else.
> 
> Cutting down on the number of terminfo definitions will make it easier to maintain these in the future.

From a try-not-to-confuse-the-user perspective, I agree with keeping the number of top-level definitions to a minimum. In terms of LOCs (and thus maintainability) it doesn't matter much though.

For example, removing the `alacritty` definition could look like this:

```diff
@@ -1,11 +1,8 @@
-alacritty|alacritty terminal emulator (X Window System),
-       use=alacritty-new,
-
 alacritty-256color|alacritty with 256 colors,
        rs1=\Ec\E]104\007, use=alacritty+256color, use=alacritty-new,
 alacritty-direct|alacritty with direct color indexing,
-       use=alacritty+indirect, use=alacritty,
+       use=alacritty+indirect, use=alacritty-new,
 alacritty-basic|modern alacritty terminal emulator - common,
        OTbs, am, bce, km, mir, msgr, xenl, AX, XT,
```

> If we want users to actually use our new terminfo, to make it possible to use features that might depend on it without unnecessary questions coming up in the issue tracker, I also think it would be a good idea to document the terminfo installation in the README.md.

That's of course a really good idea :)

**Comment by dnkl** [comment]
Btw, should the default TERM value be changed from `xterm-256color` to `alacritty-256color` (or `alacritty`, if we rename the 256-color variant)?

**Comment by chrisduerr** [comment]
> Instead of copying all definitions available for xterm, I'd recommend just copying all definitions required for Alacritty to work properly with all available features.

This was mainly talking about the top level definitions like `alacritty-256color`. The problem with providing more definitions here is that we can't just say we support 256color and direct, but have all of them in the terminfo file. We should only have the ones where we're actually willing to fix issues with. It's also very hard to remove them after they have been present, since users might have started using them since they were available, so suddenly deleting them is troublesome. This also increases the number of possible terminfo definitions which might be used by a user, so debugging issues could be more complicated.

>Btw, should the default TERM value be changed from xterm-256color to alacritty-256color (or alacritty, if we rename the 256-color variant)?

Yes, I think the default `$TERM` value should be changed to whatever we choose to pick as our primary supported terminfo file where all features Alacritty supports are enabled. This is in-line with what other terminal emulators do, like urxvt, termite or kitty.

**Comment by dnkl** [comment]
I'm all for that. In fact, I'd personally like to have a single one (the direct variant), but realize it might be taking things one step too far.

So, remove `alacritty` and keep `alacritty-256color` and `alacritty-direct`? If it's important to reduce the number of user _visible_ definitions (zsh for example, can tab-complete TERM entries, and fragments will show up there too), merge the fragments into a single base definition: `alacritty-256color` and have `alacritty-direct` inherit from it? Or do you think it's ok to keep the fragments as is?

(Note that **all** fragments present here are actually used, directly or indirectly, by `alacritty-256color` so they cannot simply be removed).

**Comment by chrisduerr** [comment]
I don't think it's important to reduce the number of supported features, it makes sense to keep everything in that case, unless there are any duplicates.

It would make sense to reduce repetition by having like an `alacritty` definition where everything basic is defined and then have `alacritty-256color` and `alacritty-direct` inherit from it. This would help when a feature like tab width needs to be tweaked, because it could be automatically changed for both term definitions.

Regarding `alacritty-256color` vs `alacritty`, I'm not quite sure what I would prefer myself. But I think having `alacritty-256color` and `alacritty-direct` would indicate that none of these are the "default" supported value and both are equally viable for usage. However having an `alacritty` term value we could make use of would make it obvious which one a beginner who has absolutely no clue what he's doing should use. Since we need to pick one for the code which Alacritty should use by default, I think I'm slightly in favor of having an `alacritty` one, but I'm very open to hear other arguments here.

**Comment by dnkl** [comment]
> It would make sense to reduce repetition by having like an alacritty definition where everything basic is defined and then have alacritty-256color and alacritty-direct inherit from it.

That's exactly how it's layed out currently, except that the base definition is split up over many small fragments. At first I thought it was a good idea to keep it that way, but now I'm starting to think it's better to merge them into one single base definition. In fact, the `alacritty` definition could probably **be** that base class, and `alacritty-direct` could inherit from it.

> I think I'm slightly in favor of having an alacritty one, but I'm very open to hear other arguments here.

I could say that someone slightly more advanced than a beginner might expect a 256color variant, because his previous terminal used `xterm-256color`. But I would just be playing the devil's advocate. I would prefer the name to be `alacritty`.

FYI, I'll be out of town over the weekend. Thus there won't be any (code) updates until next week.

**Comment by chrisduerr** [comment]
> That's exactly how it's layed out currently, except that the base definition is split up over many small fragments. At first I thought it was a good idea to keep it that way, but now I'm starting to think it's better to merge them into one single base definition. In fact, the alacritty definition could probably be that base class, and alacritty-direct could inherit from it.

Sounds good to me :+1:.

> I could say that someone slightly more advanced than a beginner might expect a 256color variant, because his previous terminal used xterm-256color. But I would just be playing the devil's advocate. I would prefer the name to be alacritty.

That's a good argument, if I'm looking for the `256color` version, I'd expect that to be the name of it. However I think that someone looking for the `256color` version should be able to figure out when the default version already is that, the terminfo also contains a short description, so something like `alacritty | Alacritty's version of xterm-256color` or similar could be used to make it clear to "advanced users" that `alacritty` is indeed the `256color` version.

> FYI, I'll be out of town over the weekend. Thus there won't be any (code) updates until next week.

This shouldn't be a problem at all, I'd expect other PRs like scrollback to be merged first anyways, to prevent conflicts in README and other touched files.

To make sure I've actually understood you correctly, here are the changes that need to be made to comply with this discussion:
 * Remove fragmentation of terminfo
 * Rename `alacritty-256color` to `alacritty`
 * Set the default `$TERM` used by Alacritty to `alacritty`
 * Remove `export TERM="alacritty"` part from README (since previous point does this)

**Comment by dnkl** [comment]
That's my understanding as well. To that list, I'd also add:

* Update documentation of `TERM` in `alacritty.yml` (and `_macos.yml`?)
* Clean up/remove capabilities `tic` complains about. These capabilities **are** taken from the xterm definitions, but `tic` still complains. I think they are related to legacy definitions or so. `tic` warns, but doesn't exit with an error. Still, it doesn't look good with all those warnings.
* Minor update to `README.md` (to the new _Terminfo_ section, added in this PR) which currently states you **must** (manually) configure `TERM` to use the *alacritty` definition.

That's it I think.

**Comment by dnkl** [comment]
WIP tag removed, as I now consider this finished.

I ended up **not** removing the capabilities tic warns about; as far as I can tell, they are used in termcap on BSD.

I also could not find any `export TERM=...` in the README (or anywhere else for that matter).

Another thing that hasn't been discussed is the connection between the key definitions in the terminfo and the key bindings in the configuration; they should match. But, since the terminfo entry is static (at least after compilation/installation), it's not possible to prevent users from changing key bindings such that they no longer match the terminfo.

Perhaps it's good enough to ensure the **default** key bindings match the terminfo?

File: alacritty.yml [hunk 1]
```diff
@@ -3,16 +3,12 @@
...
   # TERM env customization.
   #
-  # If this property is not set, alacritty will set it to xterm-256color.
-  #
-  # Note that some xterm terminfo databases don't declare support for italics.
-  # You can verify this by checking for the presence of `smso` and `sitm` in
-  # `infocmp xterm-256color`.
-  TERM: xterm-256color
+  # If this property is not set, alacritty will set it to alacritty.
+  # TERM: alacritty
```
Comment (chrisduerr) [inline alacritty.yml]: The wording here is a bit awkward and the lacking space in between the description and the commented value makes it hard to see that it's actually a yml property.

Something like this might be better:
```
# TERM variable
#
# This value is used to set the `$TERM` environment variable for
# each instance of Alacritty. If it is not present, the default value
# 'alacritty' will be used instead.
#
# TERM: alacritty
```

Comment (dnkl) [reply alacritty.yml]: No argument there. Feel free to push a fix. Otherwise I'll do it later.

File: README.md [hunk 2]
```diff
@@ -327,6 +327,32 @@ To install the completions for fish, run
...
 
+## Terminfo
+
+The terminfo database contains entries describing the terminal
+emulator's capabilities. Programs need these in order to function
+properly.
+
+Alacritty should work with the standard `xterm-256color` definition,
+but to allow programs to make best use of alacritty's capabilities,
+use it's own terminfo definition instead.
```
Comment (chrisduerr) [inline README.md]: Small typo here: It should be `its`, not `it's`.

File: README.md [hunk 3]
```diff
@@ -327,6 +327,32 @@ To install the completions for fish, run
...
+sudo tic alacritty.info
+```
+
+If you are a package maintainer, you might want to do something along
+these lines instead (`${pkgdir}` refers to the directory where
+the package is being built):
+
+```sh
+tic -o "${pkgdir}/usr/share/terminfo" alacritty.info
+```
```
Comment (chrisduerr) [inline README.md]: I don't think it's necessary to document what maintainers are supposed to do here. There should be plenty of examples and other documentation with the package manager they're using. And they might follow completely different conventions anyways.

So I think it's safe to drop this part to keep the readme a bit more minimal.

File: alacritty.info [hunk 4]
```diff
@@ -1,8 +1,122 @@
...
+	kHOM3=\E[1;3H, kHOM4=\E[1;4H, kHOM5=\E[1;5H,
+	kHOM6=\E[1;6H, kHOM7=\E[1;7H, kIC3=\E[2;3~, kIC4=\E[2;4~,
+	kIC5=\E[2;5~, kIC6=\E[2;6~, kIC7=\E[2;7~, kNXT3=\E[6;3~,
+	kNXT4=\E[6;4~, kNXT5=\E[6;5~, kNXT6=\E[6;6~,
+	kNXT7=\E[6;7~, kPRV3=\E[5;3~, kPRV4=\E[5;4~,
+	kPRV5=\E[5;5~, kPRV6=\E[5;6~, kPRV7=\E[5;7~,
+	kdch1=\E[3~,
+
+	Cr=\E]112\007, Cs=\E]12;%p1%s\007,
+	Ms=\E]52;%p1%s;%p2%s\007, Se=\E[2 q, Ss=\E[%p1%d q,
```
Comment (chrisduerr) [inline alacritty.info]: The terminfo file itself looks good to me, however there are a bunch of formatting problems. It might make sense to drop empty lines where they are not beneficial and separating obviously different blocks and make sure that indentation is always correct (looking at `npc` line).

**Comment by dnkl** [comment]
> The wording here is a bit awkward and the lacking space in between the description and the commented value makes it hard to see that it's actually a yml property.

Fixed

> Small typo here: It should be its, not it's.

Fixed

> So I think it's safe to drop this part to keep the readme a bit more minimal.

Fixed

> It might make sense to drop empty lines where they are not beneficial and separating obviously different blocks and make sure that indentation is always correct (looking at npc line).

The empty lines that are there represents the xterm fragments they came from. I think we can simply remove them all.

The `npc` was a tab-vs-spaces issue; xterm's terminfo uses tabs for indentation, while my Emacs is configured to use spaces.

The terminfo syntax doesn't require tabs, so if you're like me and prefer spaces over tabs, we could replace all tabs with spaces. For now, I've "fixed" the `npc` line by replacing the spaces with a tab.

**Comment by dnkl** [comment]
> In regards to the keyboard shortcuts, a short comment might be enough, similar to what we have added to the tabstops.

Comment added. Wording improvements are welcome :)

**Comment by dnkl** [comment]
Not sure what the norm in this project is, but my plan is to squash at least the "fix" commits before merging.

**Comment by chrisduerr** [comment]
I'm probably going to merge everything with a relatively small change like this. But you don't have to worry about that, I'll do this myself when merging.

**Comment by chrisduerr** [comment]
I've just tested it and ran into the same issue I've always ran into when trying to use a custom terminfo. When opening an application like `htop`, the mouse interaction does not work anymore when changing the `TERM` to the new `alacritty`.

Also I've noticed that this generates the `alacritty-common` as an available terminfo file, it might be nice to look into hiding an entry which is just used for reducing code duplication. I'm not sure if this is possible though.

**Comment by dnkl** [comment]
Hmm, I see the same thing with htop. Is that the only application you've had issues with? It might be that htop, and possibly others, are doing explicit string comparisons on the TERM variable, instead of looking it up in terminfo. Kind of hard to work around...

> Also I've noticed that this generates the alacritty-common as an available terminfo file, it might be nice to look into hiding an entry which is just used for reducing code duplication. I'm not sure if this is possible though.


Not directly. You can tell `tic` to only write out a specific set of entries though. So we could change the documentation to say you should run tic like this:

```sh
sudo tic -e alacritty,alacritty-direct alacritty.info
```

**Comment by dnkl** [comment]
@chrisduerr please try again. I think I found the issue. It wasn't htop specifically, but ncurses. It looks for a specific value of `kmous` and disables mouse support if it's not the expected value _unless_ `TERM` contains the substring "xterm"...

**Comment by dnkl** [comment]
> Last thing is that you forgot to add the terminfo note to the macos config file. Once that's done everything's good!

Fixed!

**Comment by papertigers** [comment]
Will we still be able to use `xterm-256color` with these changes? When I was testing out kitty I ran into some issues ssh-ing into nodes that didn't know anything about the kitty terminfo which made things difficult.  Kitty provided a way to remotely update the terminfo on the remote end but there were issues with that working for me on non OSX/Linux environments.  

Personally I would prefer to be able to keep using `xterm-256color` as alacritty currently fits my needs and works perfectly with remote systems for me.

**Comment by chrisduerr** [comment]
@papertigers There shouldn't be any issue, you can just `export TERM="xterm-256color"` and you should be good to go. You can also use the config option to permanently overwrite it to make use of the xterm terminfo.

Of course one of the reasons why we want to switch is because xterm's features are limited, so you won't have any access to new fancy stuff then. However you can still just install Alacritty's terminfo on remote systems and you should be good to go again. This will be especially easy if Alacritty ever makes it into the official repositories of the distro you're remoting into. Since then you could just install `alacritty-terminfo` and you'd be good.

File: README.md [hunk 5]
```diff
@@ -327,6 +327,24 @@ To install the completions for fish, run
...
+Alacritty should work with the standard `xterm-256color` definition,
+but to allow programs to make best use of alacritty's capabilities,
+use its own terminfo definition instead.
+
+Note that the default configuration **requires** alacritty's terminfo.
+
+To install alacritty's terminfo entry globally:
+
+```sh
+sudo tic -e alacritty,alacritty-direct alacritty.info
```
Comment (jedahan) [inline README.md]: I'm assuming most package maintainers make the decision to do this for the user or not. My question is if the supported way to install alacritty is through cargo with `cargo install alacritty`, should we have the build script do this for them? Or at the very least, post-install message 'Hey you should run this'?

Comment (jwilm) [reply README.md]: This is a really good question, and I'm not sure about the right approach. It would kind of be too bad if every time the build script ran, a new terminfo definition was installed (at least for Alacritty devs).

While `cargo install alacritty` would certainly be nice for Rust programmers, I think the long-term "supported" way to install it will be through package managers or binaries downloaded from the project. In those cases, we may have a bit more flexibility on packaging terminfo definitions.

**Comment by jwilm** [comment]
> @chrisduerr please try again. I think I found the issue. It wasn't htop specifically, but ncurses. It looks for a specific value of kmous and disables mouse support if it's not the expected value unless TERM contains the substring "xterm"...

Wow. I knew ncurses was doing something funny with string detection on $TERM value, but that is _subtle_.

Excellent work on this @dnkl. I'm particularly excited about this feature; having our very own terminfo definition which doesn't rely on xterm definitions is a big milestone for the project.

On the removal of `env.TERM = 'xterm-256color'` in the default config, this seems likely to create extra support needs on our issue tracker when new users find that their favorite applications don't work immediately. It makes sense for us to set `alacritty` from _tty.rs_, but I'm in favor of keeping xterm-256color in the config. Thoughts, @dnkl, @chrisduerr?

**Comment by chrisduerr** [comment]
That notion seems a bit odd to me. As a user I'd expect the standard configuration to reflect the default value Alacritty sets, at least this would fall in line with the rest of our configuration file. So if I would uncomment the `TERM` line and suddenly things changed, I'd think that would be pretty odd.

For existing users, nothing would change, since we wouldn't change their `TERM` value in the configuration file. So this would only affect the default behavior for people who start out with a new configuration file.

This is where `when new users find that their favorite applications don't work` comes into play. If we're making the assumption that Alacritty's terminfo file will not work properly, I don't think it's worth putting in the effort to support it. While a standalone terminfo file is always nice, I'd go on a whim here and make the bold claim that unless it's shipped by default, it's unlikely that a significant amount of users would make use of it.

I think we should strive to either make our own terminfo file work and support it as the default, or save ourselves the effort and fully rely on xterm. Note that there are things that Alacritty currently can't do by default since it uses the xterm-256color terminfo. For example nvim will not recognize it to support resetting to the default cursor. And multi selection in kakoune is another issue that didn't work because of xterm-256color terminfo.

So I'd be for trying to make this work by going all the way. Even if we supply our own terminfo, I'd be surprised if people would know it exists unless we're making use of it as our default.

As a last note I'd like to mention a trick that other terminal emulators make use of. Both termite and kitty use `xterm-termname` as their terminfo name. This would trick applications like `htop` to think that it's actually an xterm terminfo, since it queries for `xterm` anywhere in the terminfo string, not a specific terminfo name. However URxvt seems to get along even without making use of this, however it has a lot more history and thus often is considered in the same workarounds that check for `xterm` in the terminfo, so I don't think looking at URxvt for inspiration will help us much.

**Comment by jedahan** [comment]
I'm having flashbacks of querying the browser useragent string >_>

**Comment by jwilm** [comment]
@chrisduerr I'm making the assumption that most users won't bother to install our terminfo file rather than it not working properly. However, I don't think this is a great reason to not have a terminfo file that accurately reflects Alacritty's capabilities.

Eventually, we can solve this with packaging. When a user installs alacritty, the terminfo can just come with it.

Part of the problem being solved here is that we actually can't rely on xterm definitions to be correct. There's a lot of semi-recent systems that have a bad xterm terminfo which doesn't support italics, and then we get support about "why doesn't this work". At least with our own terminfo, we can point people to that.

Regarding the `xterm-termname` hack, I'd rather not. Ultimately, installing our terminfo file will fix the problem. The `htop` issue I think you're alluding to was resolved by @dnkl with the addition of the `kmous` capability to our terminfo.

In short, here's what I would like to do

1. Keep the default `TERM=alacritty` in tty.rs
2. Keep `xterm-256color` as the default in the config files; this will work on most systems. For any systems where xterm terminfo is out-of-date, we can advise to install the alacritty terminfo.
3. Merge this PR, advise people to use our terminfo but expect many won't. In general, this means being "xterm-like".
4. Once we start packaging, also provide the terminfo as part of the installation. At this point, we can remove the `xterm-256color` from the default config file.

**Comment by chrisduerr** [comment]
It's a great point that many probably wouldn't install the terminfo file themselves. I was thinking about Archlinux, wich recently got the package in the official repos and had it in the AUR for a long time, so that isn't relevant here.

I don't think this would be a great solution, but just as a potential alternative I'd like to mention that it would be possible to have automated build scripts which install the required files automatically. I personally don't really like that solution because I'm a bit allergic to applications which install stuff I don't expect them to install, but it would be a possible alternative for everyone who can't rely on their package manager to install it for them.

**Comment by jwilm** [comment]
We could potentially add a command to our CLI to install it on-demand to make it easily available.

**Comment by dnkl** [comment]
I don't think I have anything more to add to what's already been said; I believe this is mostly a packaging issue - once there are "real" packages, the alacritty terminfo *should* be the default one.

So the question is what to do until then, when most users install manually. I completely agree with @chrisduerr  that people won't be using the alacritty terminfo unless we default to it. But also understand that it's probably too much to expect everyone to perform additional installation steps.

In the end I think it comes down to what type of bug reports you guys prefer to see, and which not to...

**Comment by dnkl** [comment]
For what it's worth, I'm also against the xterm terminfo prefix.

**Comment by jedahan** [comment]
if you are installing alacritty through cargo, I think mentioning after install "you probably want to `sudo tic alacritty,alacritty.info`" is reasonable enough, and a good enough hint for package maintainers to do the same when making distro wide packages.

**Comment by jwilm** [comment]
@dnkl I've been trying to wrap my head around terminfo packaging, and it seems that **a lot** of applications manage to go the route of packaging their terminfo with _ncurses_. Is this the "accepted" way of distributing terminfo?

**Comment by dnkl** [comment]
I honestly have no idea :) I don't know if it is the applications that push their terminfos to ncurses, or if the ncurses maintainer(s) write terminfo definitions themselves.

Long term it might be a good idea. Short term it will make it difficult to synchronize feature additions to alacritty that also require terminfo updates.

**Comment by jwilm** [comment]
@dnkl I agree we don't necessarily want ours there right now, but it was a bit surprising when I went looking for how other applications distribute theirs to find this huge repository of terminfo in ncurses.

**Comment by dnkl** [comment]
AFAIK, ncurses, terminfo and xterm are fairly tightly coupled. Terminfo originally described _real_ terminals. Then came xterm. They all have the same maintainer/developer so it's not that odd.

I think most definitions are by the ncurses developers, to ensure ncurses works properly everywhere. For example, st ships their own definition, *but* there is also definitions provided by ncurses (and they are different, I think). Or take konsole; its maintainer prefers users to use the xterm(-256color) definition, even though there is a konsole definition provided by ncurses.

**Comment by jwilm** [comment]
The folks over at Kitty have apparently [investigated getting a kitty definition into ncurses](https://github.com/kovidgoyal/kitty/issues/879), but they ran into some friction and ultimately abandoned it. Ignoring the outcome of that thread, there were a number of good points around distribution from @kovidgoyal, including:

> Even if ncurses included the terminfo file today, it would be years before you could be confident of having it available everywhere

ncurses releases only periodically, and it takes time for distros to update their packages. This is spot on.

> As far as I know most distros already package kitty. If they follow the kitty packaging guidelines they should have a separate kitty-terminfo package which you can simply install on your servers

This sounds like a smart approach that we should consider when we start packaging.

One other notable item there was that Kitty went the route of using `xterm-kitty` as a TERM value, and the ncurses maintainers rejected it. I'm not informed enough to take a side in this discussion, but it's good to know that this would be a cause for rejection _if_ we ever decide to push for inclusion.

**Comment by dnkl** [comment]
> This sounds like a smart approach that we should consider when we start packaging.

Arch already does this; it has an `alacritty` and an `alacritty-terminfo` package.

**Comment by jwilm** [comment]
> I think most definitions are by the ncurses developers, to ensure ncurses works properly everywhere. For example, st ships their own definition, but there is also definitions provided by ncurses (and they are different, I think). Or take konsole; its maintainer prefers users to use the xterm(-256color) definition, even though there is a konsole definition provided by ncurses.

From what I've seen, there's a bit of both. Interesting example with konsole, though.

Strangely, the only place I could find an "official" tmux terminfo was in ncurses; tmux doesn't publish their own, at least not directly. :wink:

**Comment by jwilm** [comment]
@dnkl I missed this gem earlier

> In the end I think it comes down to what type of bug reports you guys prefer to see, and which not to...

The two types of reports I see are something like this:
1. We default to `TERM=alacritty` now. High chance of issue opening when things don't look right. We tell user to install. Later receive a ticket about the same problem because it doesn't work well on servers.
2. We default to `TERM=xterm-whatever` until packaging. Most things will work perfectly for most users. No extra work required for that majority. Minority with problems can be directed to install our terminfo.

To me, this is about _volume_ of tickets rather than type, and I see jumping the gun on `TERM=alacritty` as a potential route for a big increase in support-related tickets.

**Comment by chrisduerr** [comment]
Let me just throw in another idea here. Terminfo files are installed and it's possible to look up which terminfo files are available. It would totally be possible to try and load the `alacritty` terminfo file first and then fall back to `xterm-256color` if it is not present. This would just require a database lookup every time we start up Alacritty, which shouldn't be a massive issue.

If we ever feel comfortable that most users should have the terminfo file on their system, it would also be super easy to change the behavior and skip checking before loading. I'm not sure if other terminal emulators have done this before, but if they haven't, there might be a good reason why. Since it seems like a fairly reliable solution, even though it is a bit of a hack.

**Comment by dnkl** [comment]
@chrisduerr I assume you mean if the user hasn't configured a TERM value? I.e. the default "value" is to first try `alacritty` and then fallback to `xterm-256color`. But if the user has set a TERM value, use that as is. Correct?

Perhaps not very clean, but definitely pragmatic.

**Comment by dnkl** [comment]
@jwilm to me personally, the default value is not important, since it's mostly about supporting new users. I'll go with whatever you guys decide.

**Comment by chrisduerr** [comment]
@dnkl No, I'm talking about checking which terminfo files are actually installed on the system.

Using the `terminfo` crate this could be as simple as this:
```rust
extern crate terminfo;
use terminfo::Database;
fn main() {
    let mut term = "alacritty";
    if let Err(_) = Database::from_name("alacritty") {
        term = "xterm-256color";
    }
    println!("TERM: {}", term);
}
```

On my system this correctly resolves the terminfo if one of them doesn't exist.

**Comment by chrisduerr** [comment]
I wouldn't remove the configuration option. I'd just make sure the default is `alacritty`, if that is installed and `xterm-256color` if `alacritty` is not installed. To communicate this in the configuration file it shouldn't be a problem to just make it optional (it already is) and add the `TERM = xterm-256color` as a comment. That way we don't have to pick one by default, but we can try `alacritty` first and then fall back to `xterm-256color`, but if the config option is set, we just use that.

As I've mentioned previously I'm not sure if any other terminal emulators do something like this. Or if there's a good reason why something like this isn't a good idea. If so, please let me know. However I think it would solve all issues mentioned in this comment thread so far.

**Comment by dnkl** [comment]
Good, then we're on the same page. This is exactly what I was trying to describe in https://github.com/jwilm/alacritty/pull/1491#issuecomment-423731117

**Comment by chrisduerr** [comment]
Ah okay, sorry for the misunderstanding then. Reading it again that sounds exactly like what I'm talking about.

**Comment by jwilm** [comment]
@dnkl 

> Not too sure about that... case in point: I mostly want to use the alacritty-direct terminfo, for 24-bit support. But I really don't want it with ncmpcpp (because it crashes).

How are you setting the TERM value for that today? We don't exactly have a way to set TERM from the CLI.

@chrisduerr great idea on simply checking the system database. Users can always force something in the config if they really want to.

**Comment by chrisduerr** [comment]
> We don't exactly have a way to set TERM from the CLI.

You can just `export TERM="xterm-256color"`, right? I'm not quite sure what you mean with this comment.

**Comment by dnkl** [comment]
> How are you setting the TERM value for that today? We don't exactly have a way to set TERM from the CLI.

```sh
TERM=alacritty ncmpcpp
```

My point was that there are reasons to use different terminfo values. Hardcoding a value is thus not a good idea. Shouldn't matter now though, as it turns out it was just a misunderstanding.

> great idea on simply checking the system database. Users can always force something in the config if they really want to.

@chrisduerr @jwilm I'll go ahead and try to implement this, as well as re-word the TERM documentation in the example configuration.

**Comment by chrisduerr** [comment]
Thanks, if you have any more questions, please let me know @dnkl. Hopefully this should be simple to check. It might make sense to see how long a lookup of the database takes though, since we wouldn't want to significantly delay the startup time. You should be able to test this with `time alacritty -e true`.

**Comment by dnkl** [comment]
@chrisduerr I shamelessly copied your example and incorporated it into tty.rs. `time` didn't show any measurable differences, neither in debug nor in release (but then its resolution is pretty low).

Not sure I managed to add the `terminfo` crate dependency correctly (first time doing rust...). In particular, having to do `use self::terminfo...` seemed wrong (why `self::`?)

Looking at the code again, perhaps the `extern crate terminfo` should b moved to `lib.rs`?

File: README.md [hunk 6]
```diff
@@ -336,6 +336,24 @@ To install the completions for fish, run
...
+
+The terminfo database contains entries describing the terminal
+emulator's capabilities. Programs need these in order to function
+properly.
+
+Alacritty should work with the standard `xterm-256color` definition,
+but to allow programs to make best use of alacritty's capabilities,
+use its own terminfo definition instead.
+
+Note that the default configuration **requires** alacritty's terminfo.
```
Comment (chrisduerr) [inline README.md]: This is not accurate anymore.

File: alacritty.yml [hunk 7]
```diff
@@ -5,14 +5,14 @@
...
 # set by alacritty itself.
 env:
-  # TERM env customization
+  # TERM variable
   #
-  # If this property is not set, alacritty will set it to `xterm-256color`.
+  # This value is used to set the `$TERM` environment variable for
+  # each instance of Alacritty. If it is not present, alacritty will
+  # check the local terminfo database and use 'alacritty' if it is
+  # available, otherwise 'xterm-256color'
```
Comment (chrisduerr) [inline alacritty.yml]: This is such a nice sentence, it's a shame to end it so abruptly. I'd add a `is used.` at the end to round it off a bit nicer.

File: src/tty.rs [hunk 8]
```diff
@@ -28,6 +28,9 @@ use display::OnResize;
 use config::{Config, Shell};
 use cli::Options;
 
+extern crate terminfo;
+use self::terminfo::Database;
```
Comment (chrisduerr) [inline src/tty.rs]: The `extern crate terminfo;` line needs to be moved into `src/lib.rs`. You'll see all the other `extern crate _;` lines there, just put it next to them. Then it should also be possible to just do `use terminfo::Database`.

Just to keep things in order, I'd move the `use _` statement up below the `use libc::_;` statement. This way we have libstd imports at the top, then extern crate imports and then our own local stuff.

Comment (chrisduerr) [reply src/tty.rs]: I saw that you moved the `extern crate` already while I was reviewing the code.

The `use terminfo::Database;` still needs to be moved though.

Comment (dnkl) [reply src/tty.rs]: That, and everything else, should be fixed now.

File: README.md [hunk 9]
```diff
@@ -346,7 +346,10 @@ Alacritty should work with the standard `xterm-256color` definition,
 but to allow programs to make best use of alacritty's capabilities,
 use its own terminfo definition instead.
 
-Note that the default configuration **requires** alacritty's terminfo.
+Note: unless the user has set the `TERM` environment variable in the
```
Comment (chrisduerr) [inline README.md]: I'd personally prefer it if the `Note: ` here would just be removed. Alternatively it should be made bold and the `unless` should be capitalized, but I think the text flow is better if this paragraph just stands as another normal line in the readme rather than a highlighted note.

File: CHANGELOG.md [hunk 10]
```diff
@@ -20,6 +20,15 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
 - First click on unfocused Alacritty windows is no longer ignored on platforms other than macOS
+- Standalone terminfo definition:
+    - The alacritty terminfo definitions no longer inherit from the systems
+      `xterm` definitions.
+    - The terminfo entry `alacritty` replaces the old `alacritty` and
+      `alacritty-256color` (both were actually 256 colors).
+    - There is a new entry called `alacritty-direct` which advertises 24-bit
+      color support
+    - The default `TERM` value is no longer static; the `alacritty` entry is
+      used if available, otherwise the `xterm-256color` entry is used instead.
```
Comment (chrisduerr) [inline CHANGELOG.md]: There's a few minor changes I would propose with this changelog entry.

First of all the section about `alacritty-direct` should probably be moved to the added features. There also shouldn't be any punctuation at the end of lines. 

I'd also change the first point about not inheriting from the system's `xterm` definitions to something that shows better what the actual change is for the user (feel free to reword if you have a better idea):
`The Alacritty terminfo definition no longer requires the 'xterm' definition to be present on the system`

I'd also move the `alacritty-256color` mention to the `Removed` section. You can still mention that it is now replaced by the `alacritty` terminfo file, since that is also 256 color, but the actual effect on the user is mostly that the `alacritty-256color` terminfo definition has been removed, so if they were using it they need to change it.

Since that only leaves two points in the `Changed` section of the changelog, I'd also recommend getting rid of the `Standalone terminfo definition:` headline.

File: README.md (hunk 9 above)
Comment (dnkl) [reply README.md]: Removing `Note:` sounds good to me.

File: CHANGELOG.md (hunk 10 above)
Comment (dnkl) [reply CHANGELOG.md]: Done

**Comment by z0rc** [comment]
https://lists.gnu.org/archive/html/bug-ncurses/2018-10/msg00024.html

https://invisible-island.net/ncurses/terminfo.src.html#toc-_Alacritty
